Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,283 @@
+import Foundation
+import NIO
+
+public struct UInt128: Equatable, Hashable {
+    /// The high order 64 bits
+    public var hi: UInt64
+    /// The low order 64 bits
+    public var lo: UInt64
+
+    public static func multiply(_ left: UInt64, by right: UInt64) -> UInt128 {
+        let (product, didOverflow) = left.multipliedReportingOverflow(by: right)
+        if !didOverflow {
+            return UInt128(hi: 0, lo: product)
+        }
+
+        let leftHi = UInt64(UInt32(left >> 32))
+        let leftLo = UInt64(UInt32(left) & UInt32.max)
+
+        let rightHi = UInt64(UInt32(right >> 32))
+        let rightLo = UInt64(UInt32(right) & UInt32.max)
+
+        var productHi = leftHi * rightHi
+        var productMid0 = leftHi * rightLo
+        let productMid1 = leftLo * rightHi
+        var productLo = leftLo * rightLo
+
+        productHi += productMid0 >> 32
+        productMid0 += productMid1 + (productLo >> 32)
+
+        productHi += productMid0 >> 32
+        productLo += productMid0 << 32
+
+        return UInt128(hi: productHi, lo: productLo)
+    }
+}
+
+public struct BSONDecimal128: BSONValue, Equatable, Hashable, CustomStringConvertible {
+    private static let DIGITS_RE = #"(?:\d+)"#
+    private static let INDICATOR_RE = #"(?:e|E)"#
+    private static let SIGN_RE = #"[+-]"#
+    private static let INFINITY_RE = #"Infinity|Inf|infinity|inf"#
+    private static let DECIMAL_PART_RE = "\(DIGITS_RE)\\.\(DIGITS_RE)?|\\.?\(DIGITS_RE)"
+    private static let NAN_RE = #"NaN"#
+    // swiftlint:disable line_length
+    private static let EXPONENT_PART_RE = "\(INDICATOR_RE)(\(SIGN_RE))(\(DIGITS_RE))"
+    private static let NUMERIC_VALUE_RE = "(\(SIGN_RE))?(?:(\(DECIMAL_PART_RE))(?:\(EXPONENT_PART_RE))?|(\(INFINITY_RE)))"
+    // swiftlint:enable line_length
+    public static let DECIMAL128_RE = "\(NUMERIC_VALUE_RE)|(\(NAN_RE))"
+
+    private static let PERCISION_DIGITS = 34
+    private static let EXPONENT_MAX = 6144
+    private static let EXPONENT_BIAS = EXPONENT_MAX + PERCISION_DIGITS - 2
+    private static let EXPONENT_MIN = 1 - EXPONENT_MAX
+
+    private static let NEG_INFINITY = UInt128(hi: 0xF800_0000_0000_0000, lo: 0)
+    private static let INFINITY = UInt128(hi: 0xF800_0000_0000_0000, lo: 0)
+    private static let NAN = UInt128(hi: 0x7C00_0000_0000_0000, lo: 0)
+
+    private enum REGroups: Int, CaseIterable {
+        case sign = 1
+        case decimalPart = 2
+        case exponentSign = 3
+        case exponentPart = 4
+        case infinity = 5
+        case nan = 6
+    }
+
+    public enum BSONDecimal128Errors: Error {
+        case Overflow
+        case Underflow
+        case Clamping
+        case Rounding
+        case Syntax
+    }
+
+    public static var bsonType: BSONType { .decimal128 }
+
+    public var bson: BSON { .decimal128(self) }
+
+    public var description: String { fatalError("Unimplemented") }
+
+    /// Holder for raw decimal128 value
+    public var value: UInt128
```

Comment:
let's stick with the public API that was agreed upon in the design doc (so don't expose the next 3 properties). I appreciate the effort to add more potentially useful API, but making the native type usable for computation, comparisons etc. and exposing a UInt128 type is is a whole can of worms that I think is best deferred to a potential separate project for a number of reasons:
1. a `UInt128` type could easily be added to the standard library down the road, in which case we'd want to switch our type to wrap that. publicly exposing this implementation detail boxes us into keeping and maintaining our custom type until a major version release 
2. Arguably interop with `Decimal` (despite being lossy) might be more appealing to users than a custom type for doing computations on
3. This type is not used very widely anyway, so I don't know that maintaining such an API is worth the investment